### ISSUE TITLE
Support Apple TV 4

### ISSUE THREAD

**Issue by mrdest** [scakemyer/plugin.video.quasar#452]
Following [this](http://forum.kodi.tv/showthread.php?tid=257967&pid=2283630&highlight=atv4#pid2283630) post on the official thread, can you please compile for Darwin/ARM so users who sideloaded Kodi to Apple TV 4 will use Quasar?

**Comment by mrinsane** [comment]
Yes, we are waiting for this

**Comment by mrdest** [comment]
I tried to do it myself, but it didn't go well. What I did:

1. Cloned [cross-compiler](https://github.com/scakemyer/cross-compiler), [libtorrent-go](https://github.com/scakemyer/libtorrent-go) and [quasar](https://github.com/scakemyer/quasar)
2. Added darwin-arm64 to cross-compiler and libtorrent-go
3. Updated main Dockerfiles and travis with the new folder name.
4. Created darwin-arm64/Dockerfile in cross-compiler and libtorrent-go
5. Build cross-compiler
6. Build libtorrent-go
<failed>

**The problems I thought about:**
1. Architecture name: arm64 vs aarch64
2. Darwin version: 14 vs. 15
3. OSX version: 10.10 vs 10.11
4. OpenSSL version?
5. BOOST_VERSION?
6. SWIG_VERSION?

Would be great to get help here.

Thanks!

**Comment by mrdest** [comment]
Started uploading SDKs to: [Apple-SDKs](https://github.com/mrdest/Apple-SDKs) maybe it will be helpful to duplicate and edit:

- [cross-compiler/darwin-x64/](https://github.com/scakemyer/cross-compiler/tree/master/darwin-x64)
- [libtorrent-go/darwin-x64/](https://github.com/scakemyer/libtorrent-go/tree/master/darwin-x64)

@scakemyer, I've seen your comment in #694, and I can see why it can take a week to create binaries for iOS and tvOS, as @si458 said I can also help to compile and test on iOS and tvOS .

In my opinion, Quasar for tvOS will be a killer addon, I believe it will increase Apple's developer program registrations :)

**Comment by scakemyer** [comment]
I thought it was the same SDK but I guess I was mistaken. I hope you won't get into trouble for putting those on GitHub...

**Comment by mrdest** [comment]
No they are no the same.
I saw this repository: [MacOSX-SDKs](https://github.com/phracker/) and [cross-compiler](https://github.com/scakemyer/cross-compiler) uses it ([here](https://github.com/scakemyer/cross-compiler/blob/master/darwin-x64/Dockerfile#L19)). It should be the same idea for iOS and tvOS.

I hope too :/

**Comment by scakemyer** [comment]
It's a little more complicated than that unfortunately. It also needs [osxcross](https://github.com/tpoechtrager/osxcross/) which doesn't support iOS or tvOS, see [this issue](https://github.com/tpoechtrager/osxcross/issues/38) which has a few useful links about setting up the toolchain for iOS; not sure how or if it could apply to tvOS.

**Comment by mrdest** [comment]
I will look into it. Thanks for checking!

**Comment by elgatito** [comment]
There is an experimental branch of migrating from libtorrent to gotorrent ( anacrolix/torrent , pure go-language torrent library) - https://github.com/elgatito/quasar/tree/gotorrent 

I could not build it with existing docker images. 
Maybe someone can try to build it.

**Comment by ghost** [comment]
https://github.com/PopcornTimeTV/PopcornTimeTV/releases
https://github.com/PopcornTimeTV/PopcornTorrent
This app seems to work okay on my Apple TV 4, and apparently it uses libtorrent, could someone use that code somehow for Quasar?

**Comment by elgatito** [comment]
If someone wants to test on iOS, you can try my fork https://github.com/elgatito/plugin.video.elementum . I want to get some feedback from iOS users.

**Comment by mrdest** [comment]
@elgatito Does it support tvOS?

**Comment by elgatito** [comment]
@mrdest , you can try download and run the biggest zip and share kodi.log after all. 
there is a build for ios and osx, tvOS is another os from Apple?

**Comment by mrdest** [comment]
I'll try. Yes, tvOS is for Apple TV 4 and 4K.
Kodi builds: http://mirrors.kodi.tv/test-builds/darwin/tvos/
Instructions to sideload Kodi for Apple TV: http://www.redmondpie.com/how-to-sideload-install-kodi-on-apple-tv-4/

**Comment by mrdest** [comment]
[kodi.log](http://bit.ly/2lWA8t7)
23:22:48.603 T:6210416640  NOTICE: [plugin.video.elementum] elementumd: starting elementumd
23:22:48.604 T:6210416640  NOTICE: [plugin.video.elementum] Kodi's JSON-RPC service is available, starting up...
23:22:48.605 T:6210416640  NOTICE: [plugin.video.elementum] Get Binary detection premature: /var/mobile/Containers/Data/Application/4B9FDBC1-5F56-4FEE-AFF4-9284EFF654AD/Library/Caches/home/userdata/addon_data/plugin.video.elementum/bin/_x64, /var/mobile/Containers/Data/Application/4B9FDBC1-5F56-4FEE-AFF4-9284EFF654AD/Library/Caches/home/userdata/addon_data/plugin.video.elementum/bin/_x64/elementum
23:22:48.607 T:6210416640  NOTICE: [plugin.video.elementum] System information: _x64 17.2.0
23:22:48.608 T:6210416640  NOTICE: [plugin.video.elementum] Kodi build version: 17.5 Git:2017-10-24-b184048-dirty
23:22:48.608 T:6210416640  NOTICE: [plugin.video.elementum] OS type: Darwin
23:22:48.608 T:6183923712   DEBUG: JSONRPC Server: New connection detected
23:22:48.609 T:6210416640  NOTICE: [plugin.video.elementum] uname: ('Darwin', 'Apple-TV', '17.2.0', 'Darwin Kernel Version 17.2.0: Fri Sep 29 17:54:22 PDT 2017; root:xnu-4570.20.62~1/RELEASE_ARM64_T7000', 'AppleTV5,3', '')
23:22:48.609 T:6183923712    INFO: JSONRPC Server: New connection added
23:22:48.609 T:6210416640  NOTICE: [plugin.video.elementum] Binary dir: False, item: False

**Comment by elgatito** [comment]
Yes, OS detection failed. You can try to manually run different binaries, to see how it loads. So far i've read, tvos is different from ios.

**Comment by mrdest** [comment]
Tested also on iPad. Installation was successful but restarting Kodi shows this message:

> No available JSON_RPC connection to Kodi.
> Enable both Application control options in Service settings/Control and restart Kodi.

All options are checked (enabled).
[kodi.log](http://bit.ly/2zj1O0G)

> 00:11:02.215 T:6126841856 WARNING: JSONRPC: Could not parse type "Setting.Details.SettingList"
> ..
> 00:11:24.715 T:6126841856  NOTICE: CWebServer[8080]: Started
> 00:11:26.745 T:6126841856  NOTICE: starting upnp client
> 00:11:27.240 T:6126841856  NOTICE: starting upnp server
> 00:11:27.250 T:6126841856  NOTICE: stopping upnp client
> 00:11:27.410 T:6126841856  NOTICE: starting upnp client
> 00:11:28.040 T:6126841856  NOTICE: starting upnp renderer
> 00:11:29.554 T:6139359232   ERROR: GetDirectory - Error getting /private/var/mobile/Containers/Data/Application/8C119937-6A64-4A6D-85B2-0D66A9D56DE3/Documents/Inbox/
> 00:11:29.554 T:6139359232   ERROR: GetDirectory - Error getting special://envhome/Documents/Inbox/
> 00:11:35.969 T:6139359232   ERROR: GetDirectory - Error getting /private/var/mobile/Containers/Data/Application/8C119937-6A64-4A6D-85B2-0D66A9D56DE3/Documents/Inbox/
> 00:11:35.969 T:6139359232   ERROR: GetDirectory - Error getting special://envhome/Documents/Inbox/
> 00:12:01.070 T:6130282496 WARNING: DARWINIOS: underflow (0 vs 8192 bytes)

Hope this helps.

**Comment by elgatito** [comment]
Have oyh enabled remote controls in kodi settings?

**Comment by mrdest** [comment]
Yes I did.

**Comment by mrdest** [comment]
> Unable to find Elementum binary <br>Required platform: ios_arm

I think the problem is [cross-compiler](https://github.com/elgatito/cross-compiler) does not have task for ios_arm*. The [release page](https://github.com/elgatito/plugin.video.elementum/releases) has the builds, though.
Related: #445

**Comment by elgatito** [comment]
Yes, there is no ios_arm arch, only ios_arm-7 and ios-arm64. Can you try copy directory "resources/bin/ios_arm-7/" info "resources/bin/ios_arm/" and try then? Just to see what happen when binary is loaded.

**Comment by mrdest** [comment]
> ...AttributeError("'module' object has no....

I can't see the full notification and it's not written to the kodi.log file.
